### DISCUSSION TITLE
LDM Notes for April 6th, 2022

### DISCUSSION THREAD

**Discussion by MadsTorgersen** [dotnet/csharplang#5996] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-04-06.md

## Agenda

1. Unresolved questions for static virtual members
2. Parameter null checking

**Comment by orthoxerox** [comment]
> [](https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-04-06.md#4-use-a-different-syntax)Generating runtime checks for every nonnullable parameter type may be overkill because you get checking in too many places, which is a performance hit, but also may not always be what you want. Of course that could be mitigated by an opt-out syntax on parameters, but now you're back to having just as much syntax, and sort of an arms race between opt-in and opt-out mechanisms.

I can see two groups of people that want runtime checks:

- people whose libraries are consumed from null-oblivious contexts (aka people working with established codebases)
- people whose libraries are consumed by people who ignore warnings (aka everyone that releases anything on NuGet)

For both groups injecting runtime checks into every public method might actually be what they want. However, *consuming* null-oblivious libraries requires explicit null checks of the return values, and I think we still need a simple syntax for that. Should it be reused for parameters? Honestly, I don't think so. `!!` after a method call says, "I don't care about what happens on the sad path, this shouldn't even happen". `!!` after a parameter says, "My method is watertight". I want to grep for these things separately during a code review.
